worlds/tunic/combat_logic.py

```diff
@@ -0,0 +1,422 @@
+from worlds.AutoWorld import LogicMixin
+
+
+# the vanilla stats you are expected to have to get through an area, based on where they are in vanilla
+class AreaStats(NamedTuple):
+    att_level: int
+    def_level: int
+    potion_level: int  # all 3 are before your first bonfire after getting the upgrade page, third costs 1k
+    hp_level: int
+    sp_level: int
+    mp_level: int
+    potion_count: int
+    equipment: List[str] = []
+    is_boss: bool = False
+
+
+area_data: Dict[str, AreaStats] = {
+    # The upgrade page is right by the Well entrance. Upper Overworld by the chest in the top right might need something
+    "Overworld": AreaStats(1, 1, 1, 1, 1, 1, 0, ["Stick"]),
+    "East Forest": AreaStats(1, 1, 1, 1, 1, 1, 0, ["Sword"]),
+    "Before Well": AreaStats(1, 1, 1, 1, 1, 1, 3, ["Sword", "Shield"]),
+    # learn how to upgrade
+    "Beneath the Well": AreaStats(2, 1, 3, 3, 1, 1, 3, ["Sword", "Shield"]),
+    "Dark Tomb": AreaStats(2, 2, 3, 3, 1, 1, 3, ["Sword", "Shield"]),
+    "West Garden": AreaStats(2, 3, 3, 3, 1, 1, 4, ["Sword", "Shield"]),
+    "Garden Knight": AreaStats(3, 3, 3, 3, 2, 1, 4, ["Sword", "Shield"], is_boss=True),
+    # get the wand here
+    "Beneath the Vault": AreaStats(3, 3, 3, 3, 2, 1, 4, ["Sword", "Shield", "Magic"]),
+    "Eastern Vault Fortress": AreaStats(3, 3, 3, 4, 3, 2, 4, ["Sword", "Shield", "Magic"]),
+    "Siege Engine": AreaStats(3, 3, 3, 4, 3, 2, 4, ["Sword", "Shield", "Magic"], is_boss=True),
+    "Frog's Domain": AreaStats(3, 4, 3, 5, 3, 3, 4, ["Sword", "Shield", "Magic"]),
+    # the second half of Atoll is the part you need the stats for, so putting it after frogs
+    "Ruined Atoll": AreaStats(4, 4, 3, 5, 3, 3, 5, ["Sword", "Shield", "Magic"]),
+    "The Librarian": AreaStats(4, 4, 3, 5, 3, 3, 5, ["Sword", "Shield", "Magic"], is_boss=True),
+    "Quarry": AreaStats(5, 4, 3, 5, 3, 3, 5, ["Sword", "Shield", "Magic"]),
+    "Rooted Ziggurat": AreaStats(5, 5, 3, 5, 3, 3, 6, ["Sword", "Shield", "Magic"]),
+    "Boss Scavenger": AreaStats(5, 5, 3, 5, 3, 3, 6, ["Sword", "Shield", "Magic"], is_boss=True),
+    "Swamp": AreaStats(1, 1, 1, 1, 1, 1, 6, ["Sword", "Shield", "Magic"]),
+    "Cathedral": AreaStats(1, 1, 1, 1, 1, 1, 6, ["Sword", "Shield", "Magic"]),
+    # marked as boss because the garden knights can't get hurt by stick
+    "Gauntlet": AreaStats(1, 1, 1, 1, 1, 1, 6, ["Sword", "Shield", "Magic"], is_boss=True),
+    "The Heir": AreaStats(5, 5, 3, 5, 3, 3, 6, ["Sword", "Shield", "Magic", "Laurels"], is_boss=True),
+}
+
+
+# these are used for caching which areas can currently be reached in state
+boss_areas: List[str] = [name for name, data in area_data.items() if data.is_boss and name != "Gauntlet"]
+non_boss_areas: List[str] = [name for name, data in area_data.items() if not data.is_boss]
+
+
+class CombatState(IntEnum):
+    unchecked = 0
+    failed = 1
+    succeeded = 2
+
+
+def has_combat_reqs(area_name: str, state: CollectionState, player: int) -> bool:
```

Comment:
This is pretty slick, it is big enough I'd consider having a couple test around it just to make debugging easier as well as clarifying the expected behavior, obviously not a blocker but wanted to raise